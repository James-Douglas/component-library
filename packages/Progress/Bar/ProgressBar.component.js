import React from 'react';
import PropTypes from 'prop-types';
import { ManorProvider } from '@comparethemarketau/manor-provider';
import { StyledBar, StyledProgress } from './ProgressBar.styles';

const Bar = ({
  value,
  isSticky,
  stuck,
  theme,
}) => (
  <ManorProvider theme={theme}>
    <StyledBar stuck={stuck} isSticky={isSticky} value={value}>
      <StyledProgress max="100" value={value} />
    </StyledBar>
  </ManorProvider>
);

Bar.propTypes = {
  /**
   * Percentage completed (0 - 100)
   */
  value: PropTypes.number.isRequired,
  /**
   * Determines if the bar should be sticky
   */
  isSticky: PropTypes.bool,
  /**
   * Determines if the bar is in the stuck state
   */
  stuck: PropTypes.bool,
  // eslint-disable-next-line react/forbid-prop-types
  /**
   * Manor theme, if not provided the ctm theme will be used.
   */
  // eslint-disable-next-line react/forbid-prop-types
  theme: PropTypes.object,
};

Bar.defaultProps = {
  isSticky: false,
  stuck: false,
  theme: undefined,
};

export default Bar;
