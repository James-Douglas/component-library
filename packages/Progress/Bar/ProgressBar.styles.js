import styled, { css } from 'styled-components';

export const StyledBar = styled.div`
  width: 100%;
  height: 0.6rem;
  font-size: ${({ theme }) => theme.fontSize.sm};
  z-index: ${({ theme }) => (theme.zIndex[30])};
  position: ${({ isSticky, stuck }) => (isSticky || stuck ? 'fixed' : 'relative')};
  top: ${({ isSticky, stuck }) => (isSticky || stuck ? '0' : 'inherit')};
`;

export const sharedStyleProgress = css`
  background: ${({ theme }) => theme.progress.bar.progressBackground};
  transition: width 0.4s ease-in-out;
  border-radius:  ${({ theme, value }) => (value.toString() === '100' ? theme.borderRadius.none : `0 ${theme.borderRadius.full} ${theme.borderRadius.full} 0`)};
`;

export const StyledProgress = styled.progress`
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  /* IE-11 styling */
  color: ${({ theme }) => theme.progress.bar.progressBackground};
  &[value]  {
    /* Reset the default appearance */
    -webkit-appearance: none;
    -moz-appearance: none;
    appearance: none;
    /* Get rid of default border in IE. */
    border: none;
  }
  
  ::-webkit-progress-bar {
    background: ${({ theme }) => theme.progress.bar.backgroundValue};
  }
  ::-webkit-progress-value {
    ${sharedStyleProgress}
  }
  /* ie11 & edge remove black right border */
  &::-ms-fill {
    border-color: currentColor;
  }
  ::-moz-progress-bar {
    ${sharedStyleProgress}
  }
`;
