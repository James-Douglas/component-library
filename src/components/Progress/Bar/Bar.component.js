import React from 'react';
import PropTypes from 'prop-types';
import styles from '../Tracker/styles';

const Bar = ({
  value,
  backwards,
  isSticky,
  stuck,
}) => {
  const stickyClass = isSticky ? 'sticky' : '';
  const stuckClass = stuck ? 'stuck' : '';
  return (
    <div className={`progress-container ${stickyClass} ${stuckClass}`}>
      <style jsx>{styles}</style>
      <progress className={`progress ${(value === 100) ? 'complete' : ''}`} max="100" value={value} />
      <div className={`label  ${(!backwards) ? 'forwards' : ''}`} style={{ marginLeft: `${value}vw`, right: value > 10 ? '4.8rem' : '3.5rem' }}>
        {`${value}%`}
      </div>
    </div>
  );
};


Bar.propTypes = {
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  backwards: PropTypes.bool,
  isSticky: PropTypes.bool,
  stuck: PropTypes.bool,
};

Bar.defaultProps = {
  value: 70,
  backwards: false,
  isSticky: false,
  stuck: false,
};

export default Bar;
