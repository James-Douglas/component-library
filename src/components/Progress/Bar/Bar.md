
# Progress Bar

By passing an array of steps, links we can visually represent the current progress of journeys. Steps and links can also have active and disabled states.

## Props

* value (number or string) - the amount of progress bar to fill, as a percentage
## Usage

### Non-sticky Bar
~~~js
<Bar value={50} />

~~~

~~~js
<Bar value='50' />

~~~



### Sticky Bar
~~~js
<StickyBar value={50} />

~~~

~~~js
<StickyBar value='50' />

~~~