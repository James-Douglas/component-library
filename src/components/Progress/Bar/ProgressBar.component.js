import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const StyledBar = styled.div`
  width: 100%;
  height: 0.6rem;
  font-size: ${({ theme }) => theme.fontSize.sm};
  z-index: ${({ theme }) => (theme.zIndex[30])};
  position: ${({ isSticky, stuck }) => (isSticky || stuck ? 'fixed' : 'relative')};
  top: ${({ isSticky, stuck }) => (isSticky || stuck ? '0' : 'inherit')};
`;

const sharedStyleProgress = css`
  background: ${({ theme }) => theme.progress.bar.progressBackground};
  transition: width 0.4s ease-in-out;
  border-radius:  ${({ theme, value }) => (value.toString() === '100' ? theme.borderRadius.none : `0 ${theme.borderRadius.full} ${theme.borderRadius.full} 0`)};
`;

const StyledProgress = styled.progress`
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  /* IE-11 styling */
  color: ${({ theme }) => theme.progress.bar.progressBackground};
  &[value]  {
    /* Reset the default appearance */
    -webkit-appearance: none;
    -moz-appearance: none;
    appearance: none;
    /* Get rid of default border in IE. */
    border: none;
  }
  
  ::-webkit-progress-bar {
    background: ${({ theme }) => theme.progress.bar.backgroundValue};
  }
  ::-webkit-progress-value {
    ${sharedStyleProgress}
  }

  /* ie11 & edge remove black right border */
  &::-ms-fill {
    border-color: currentColor;
  }

  ::-moz-progress-bar {
    ${sharedStyleProgress}
  }
`;

const Bar = ({
  value,
  isSticky,
  stuck,
}) => (
  <StyledBar stuck={stuck} isSticky={isSticky} value={value}>
    <StyledProgress max="100" value={value} />
  </StyledBar>
);


Bar.propTypes = {
  /**
   * Percentage completed (0 - 100)
   */
  value: PropTypes.number.isRequired,
  /**
   * Determines if the bar should be sticky
   */
  isSticky: PropTypes.bool,
  /**
   * Determines if the bar is in the stuck state
   */
  stuck: PropTypes.bool,
};

Bar.defaultProps = {
  isSticky: false,
  stuck: false,
};

export default Bar;
