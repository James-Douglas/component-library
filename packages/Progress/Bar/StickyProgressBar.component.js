import React from 'react';
import PropTypes from 'prop-types';
import { useIsSticky } from '@comparethemarketau/manor-hooks';
import Bar from './ProgressBar.component';

const StickyBar = ({
  value, theme,
}) => {
  const isStuck = useIsSticky(60);
  return (
    <Bar value={value} isSticky={isStuck} stuck={isStuck} theme={theme} />
  );
};

StickyBar.propTypes = {
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  // eslint-disable-next-line react/forbid-prop-types
  /**
   * Manor theme, if not provided the ctm theme will be used.
   */
  // eslint-disable-next-line react/forbid-prop-types
  theme: PropTypes.object,
};

StickyBar.defaultProps = {
  value: 70,
  theme: undefined,
};

export default StickyBar;
