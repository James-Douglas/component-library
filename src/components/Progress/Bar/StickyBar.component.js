import React from 'react';
import PropTypes from 'prop-types';
import useIsSticky from '../../../hooks/useIsSticky';
import Bar from './Bar.component';

const StickyBar = ({
  value,
}) => {
  const isStuck = useIsSticky(60);
  return (
    <Bar value={value} isSticky={isStuck} stuck={isStuck} />
  );
};

StickyBar.propTypes = {
  value: PropTypes.string,
};

StickyBar.defaultProps = {
  value: '70',
};

export default StickyBar;
