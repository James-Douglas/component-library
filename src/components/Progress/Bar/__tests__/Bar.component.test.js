import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';
import Bar from '../Bar.component';

describe('Bar', () => {
  it('renders correctly with value as string', () => {
    const { getByText } = render(<Bar value="60" />);
    expect(getByText('60%')).toBeInTheDocument();
  });
  it('renders correctly with value as number', () => {
    const { getByText } = render(<Bar value={55} />);
    expect(getByText('55%')).toBeInTheDocument();
  });
  it('should check scroll and add isSticky and stuck props', () => {
    const { container } = render(<Bar value="60" isSticky stuck />);
    expect(container.firstChild).toHaveStyleRule('position', 'fixed');
  });
  it('should check progress value', () => {
    const { container } = render(<Bar value={100} />);
    const list = container.getElementsByTagName('progress')[0];
    expect(list).toHaveAttribute('value', '100');
  });
  it('should check label styles', () => {
    const { getByText } = render(<Bar value="60" />);
    expect(getByText('60%')).toHaveStyleRule('right', '4.8rem');
  });
  it('should check styles', () => {
    const { getByText } = render(<Bar value="60" />);
    expect(getByText('60%')).toHaveStyle(`
     margin-left: 60vw;
     right: 4.8rem;
   `);
  });
  it('should check styles for value less than 10', () => {
    const { getByText } = render(<Bar value="6" />);
    expect(getByText('6%')).toHaveStyle(`
      margin-left: 6vw;
      right: 3.5rem;
    `);
  });
});
