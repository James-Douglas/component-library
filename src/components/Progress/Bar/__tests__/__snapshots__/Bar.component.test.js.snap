// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar renders correctly with value as number 1`] = `"<div class=\\"progress-container  \\"><style></style><progress class=\\"progress \\" max=\\"100\\" value=\\"55\\"></progress><div class=\\"label  forwards\\" style=\\"margin-left: 55vw; right: 4.8rem;\\">55%</div></div>"`;

exports[`Bar renders correctly with value as string 1`] = `"<div class=\\"progress-container  \\"><style></style><progress class=\\"progress \\" max=\\"100\\" value=\\"66\\"></progress><div class=\\"label  forwards\\" style=\\"margin-left: 66vw; right: 4.8rem;\\">66%</div></div>"`;

exports[`Bar should check scroll and update isSticky and stuck props 1`] = `"<div class=\\"progress-container sticky stuck\\"><style></style><progress class=\\"progress \\" max=\\"100\\" value=\\"70\\"></progress><div class=\\"label  forwards\\" style=\\"margin-left: 70vw; right: 4.8rem;\\">70%</div></div>"`;

exports[`Bar should check scroll and update isSticky and stuck props 2`] = `"<div class=\\"progress-container  \\"><style></style><progress class=\\"progress \\" max=\\"100\\" value=\\"70\\"></progress><div class=\\"label  forwards\\" style=\\"margin-left: 70vw; right: 4.8rem;\\">70%</div></div>"`;
